packages/babel-traverse/src/context.js

```diff
@@ -120,8 +120,9 @@ export default class TraversalContext {
       }
 
       // ensure we don't visit the same node twice
-      if (visited.indexOf(path.node) >= 0) continue;
-      visited.push(path.node);
+      const { node } = path;
+      if (visited.has(node)) continue;
+      if (node) visited.add(node);
```

Comment:
Note that when `path.node` is `null`, the current behaviour is to visit this path once but skip other nullish `node` paths since `null` can be hold in array. In this PR they will all be visited. This could be a breaking change in a way but if two paths whose `path.node` is null, I think their equality is not well-defined.

The usage of `WeakSet` is intentional, so we don't leak memory when `path.node` is GCed.